src/python/WMCore/MicroService/MSTransferor/MSTransferor.py

```diff
@@ -352,221 +330,80 @@ def checkDataLocation(self, wflow):
-                            msg += "belongs to the campaign: %s, which does not match the campaign SiteWhiteList: %s "
-                            msg += "and SiteBlackList: %s. While the workflow has been assigned to: %s"
-                            self.logger.error(msg, wflow.getName(), dsetName, dataIn['campaign'],
-                                              campConfig['SiteWhiteList'], campConfig['SiteBlackList'], psns)
-                if not commonPsns:
-                    # returns an empty set, which will make this workflow to be skipped for the moment
-                    return commonPsns
-
-        pnns = self._getPNNsFromPSNs(commonPsns)
-        self.logger.info("  found a PSN list: %s, which maps to a list of PNNs: %s", commonPsns, pnns)
-        return pnns
+                    campSecLocations = campConfig['Secondaries'].get(dsetName, [])
+                campBasedLocation.append(set(campSecLocations))
+
+                if not campSecLocations:
+                    msg = "Workflow has been incorrectly assigned: %s. The secondary dataset: %s, "
+                    msg += "belongs to the campaign: %s, with does not define the secondary "
+                    msg += "dataset or it has defined an empty location list."
+                    self.logger.error(msg, wflow.getName(), dsetName, dataIn['campaign'])
+                    return
+
+                # compare the expected locations against the current availability
+                missingDestinations = list(set(campSecLocations) - set(secLocation))
+                msg = "it has secondary pileup: %s, with a total size of: %s GB, "
+                msg += "currently at: %s, campaign expected at: %s and missing replicas at: %s"
+                self.logger.info(msg, dsetName, gigaBytes(secSize), secLocation,
+                                 campSecLocations, missingDestinations)
+                if missingDestinations:
+                    # then update this pileup location to get rule(s) on it
+                    self.logger.info("pileup %s will get container rules on: %s", dsetName, missingDestinations)
+                    pileupInput[dsetName]['locations'] = missingDestinations
+                else:
+                    self.logger.info("pileup %s already available at the expected locations", dsetName)
+                    # then remove it from the samples to get a data placement rule
+                    pileupInput.pop(dsetName)
+
+        # consider the workflow sitelist for this final location
+        if len(campBasedLocation) == 1:
+            # then there is only one pileup dataset in the workflow
+            wflowFinalLocation = campBasedLocation[0] & wflowPnns
+        else:
+            # then there are multiple pileups, get an intersection of their
+            # expected location as final workflow destination
+            wflowFinalLocation = campBasedLocation[0].intersection(*campBasedLocation)
```

Comment:
I'm not sure I understand logic of operation here, i.e. you are taking first element of `campBasedLocation` and intersect it with the entire list. Why first? I think you should explain the operation in a comment above by clarifying what is stored in a first element and why intersection with entire list is required, e.g. the first element of `campBasedLocation ` is original pileup, the rest are additional pileups and need to intersect them together.